### PR TITLE
feat(cli): convert JSON reports into a different format

### PR DESCRIPTION
**Pull request by knqyf263**
## Description
This PR adds a new subcommand to convert JSON reports into a different format.

### Usage

```
$ trivy image -f json -o result.json --list-all-pkgs alpine:3.17
$ trivy convert --format cyclonedx --output result.cdx result.json
2023-05-23T10:24:46.425+0300    INFO    "--format cyclonedx" disables security scanning. Specify "--scanners vuln" explicitly if you want to include vulnerabilities in the CycloneDX report.
```

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/3243

## Related PRs
- https://github.com/aquasecurity/trivy/pull/3014

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [ ] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by utix** [comment]
Hi,
if I can help ping me

File: pkg/commands/convert/run.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		return xerrors.Errorf("file open error: %w", err)
+	}
+	defer f.Close()
+
+	var r types.Report
+	if err = json.NewDecoder(f).Decode(&r); err != nil {
+		return xerrors.Errorf("json decode error: %w", err)
+	}
+
+	if err = result.Filter(ctx, r, opts.FilterOpts()); err != nil {
```
Comment (DmitriyLewen) [inline pkg/commands/convert/run.go:30]: we need exclude some result from report.
e.g. secrets/misconfigs for spdx report:
```
➜  ls -hl 
-rw-rw-r-- 1 dmitriy dmitriy  891 авг  3  2019 server.key

➜  ./trivy -d fs -f json -o report.json .    
...
2023-05-24T10:19:50.094+0600	INFO	Number of language-specific files: 0
2023-05-24T10:19:50.094+0600	DEBUG	Secret file: server.key
➜   ./trivy -d convert -f spdx-json report.json   
2023-05-24T10:20:09.315+0600	DEBUG	["cyclonedx" "spdx" "spdx-json" "github"] automatically enables '--list-all-pkgs'.
2023-05-24T10:20:09.315+0600	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-05-24T10:20:09.315+0600	INFO	"--format spdx" and "--format spdx-json" disable security scanning
2023-05-24T10:20:09.315+0600	DEBUG	Writing report to output...
2023-05-24T10:20:09.316+0600	FATAL	unable to write results:
    github.com/aquasecurity/trivy/pkg/commands/convert.Run
        /home/dmitriy/work/aquasecurity/trivy/pkg/commands/convert/run.go:36
  - failed to write results:
    github.com/aquasecurity/trivy/pkg/report.Write
        /home/dmitriy/work/aquasecurity/trivy/pkg/report/writer.go:137
  - failed to save spdx json:
    github.com/aquasecurity/trivy/pkg/report/spdx.Writer.Write
        /home/dmitriy/work/aquasecurity/trivy/pkg/report/spdx/spdx.go:39
  - json: error calling MarshalJSON for type common.DocElementID: failed to marshal empty DocElementID
```

**Review by DmitriyLewen — CHANGES_REQUESTED** [review]
I found 1 bug. @knqyf263 take a look please.

The rest seems good to me.

File: docs/docs/references/configuration/cli/trivy_convert.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+```
+trivy convert [flags] RESULT_JSON
+```
+
+### Examples
+
+```
+  # report conversion
+  $ trivy image --format json --output result.json --list-all-pkgs debian:11
+  $ trivy convert --format cyclonedx --output result.cdx result.json
```
Comment (chen-keinan) [inline docs/docs/references/configuration/cli/trivy_convert.md:14]: will it work for output produced from trivy `k8s sub-command` ? maybe it should not
when tested it I was getting empty bom:
`trivy k8s cluster --format json --output result.json`
`trivy convert --format cyclonedx --output result.cdx result.json`

```json
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "serialNumber": "urn:uuid:c0f5a397-0145-4585-9931-373b4b0e7b3f",
  "version": 1,
  "metadata": {
    "timestamp": "2023-05-24T05:46:02+00:00",
    "tools": [
      {
        "vendor": "aquasecurity",
        "name": "trivy",
        "version": "dev"
      }
    ],
    "component": {
      "type": "",
      "name": "",
      "properties": [
        {
          "name": "aquasecurity:trivy:SchemaVersion",
          "value": "0"
        }
      ]
    }
  },
  "components": [],
  "dependencies": [
    {
      "ref": "",
      "dependsOn": []
    }
  ],
  "vulnerabilities": []
}
```

Comment (knqyf263) [reply docs/docs/references/configuration/cli/trivy_convert.md:14]: Yeah, nice catch. I forgot to add notes that k8s and aws results don't work.

File: pkg/commands/convert/run.go (hunk 1 above)
Comment (knqyf263) [reply pkg/commands/convert/run.go:30]: How did you reproduce it? It works in my env.

Comment (knqyf263) [reply pkg/commands/convert/run.go:30]: Oh, `-f spdx` works, but `-f spdx-json` doesn't. I'll look into it. Thanks.

Comment (DmitriyLewen) [reply pkg/commands/convert/run.go:30]:
```
➜  4452 cat Dockerfile
FROM alpine

➜  4452 ./trivy -d config -f json -o report.json ./Dockerfile
2023-05-24T13:11:13.664+0600	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-05-24T13:11:13.667+0600	DEBUG	cache dir:  /home/dmitriy/.cache/trivy
2023-05-24T13:11:13.667+0600	INFO	Misconfiguration scanning is enabled
2023-05-24T13:11:13.668+0600	DEBUG	Policies successfully loaded from disk
2023-05-24T13:11:13.689+0600	DEBUG	Walk the file tree rooted at 'Dockerfile' in parallel
2023-05-24T13:11:13.689+0600	DEBUG	Scanning Dockerfile files for misconfigurations...
2023-05-24T13:11:14.194+0600	DEBUG	OS is not detected.
2023-05-24T13:11:14.194+0600	INFO	Detected config files: 1
2023-05-24T13:11:14.194+0600	DEBUG	Scanned config file: Dockerfile
➜  4452 ./trivy convert -f spdx-json report.json
2023-05-24T13:11:38.856+0600	INFO	"--format spdx" and "--format spdx-json" disable security scanning
2023-05-24T13:11:38.857+0600	FATAL	unable to write results: failed to write results: failed to save spdx json: json: error calling MarshalJSON for type common.DocElementID: failed to marshal empty DocElementID

```

File: docs/docs/configuration/reporting.md [hunk 3]
```diff
@@ -353,9 +354,31 @@ The following example shows use of default HTML template when Trivy is installed
 $ trivy image --format template --template "@/usr/local/share/trivy/templates/html.tpl" -o report.html golang:1.12-alpine
 ```
 
-## SBOM
+### SBOM
 See [here](../supply-chain/sbom.md) for details.
 
+## Converting
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply docs/docs/configuration/reporting.md:360]: This command is not for scanning targets. I don't have an idea of the best place to put this section. Any feedback is welcome. Instead, I added Q&A.
https://github.com/aquasecurity/trivy/pull/4452/files#diff-0c85323615711ebddb9909811666e5017d2f4dc04afe9e2e93abde824aa99345R7-R8

File: pkg/commands/convert/run.go (hunk 1 above)
Comment (knqyf263) [reply pkg/commands/convert/run.go:30]: Fixed in https://github.com/aquasecurity/trivy/pull/4452/commits/73fa6bdcf7c940954039d2b3f662021d79e485fd

**Review by DmitriyLewen — APPROVED** [review]

**Comment by knqyf263** [comment]
@utix Thanks for your idea!

**Comment by itaysk** [comment]
The command is converting from json to anything else, and I'm a little concerned that json is ambiguous description for this, since it doesn't say what data is expected. for example, if someone created a custom json output with go-template, or spdx-json, or compliance output, obviously the command will not work in these cases. I think we need to call this "Trivy full json" or something similar, instead of "json" which can mean many things. For sure I'd clarify this in the docs, up to you if it needs to reflect in the CLI.

**Comment by knqyf263** [comment]
The CLI says "Trivy JSON". IMO, it is clear enough. Do you want to change it to "Trivy full JSON"?
```
$ trivy -h | grep convert
  convert     Convert Trivy JSON report into a different format
```

**Pull request merged**